src/ext/respToTcpLoader.c

```diff
@@ -9,72 +9,125 @@
 
     RdbParser *p;
     int fd;
 };
 
 /* return 0 for success. 1 Otherwise. */
-int readReplies(RdbxRespToTcpLoader *ctx, int numToRead) {
-    char buffer[2048];
-    int numRead = 0;
-
-    while (numRead < numToRead) {
-        /* Read a single-line reply into the buffer */
-        ssize_t bytesRead = read(ctx->fd, buffer, sizeof(buffer) - 1);
-
-        /* TODO: Check responses for errors. */
-        /* TODO: Record pending commands in case of failure */
-        if (bytesRead > 0) {
-            /* Iterate through the read bytes to find the end of the lines */
-            for (int i = 0; i < bytesRead; i++) {
-                if (buffer[i] == '\n') {
-                    numRead++;
-                }
+static int readReplies(RdbxRespToTcpLoader *ctx, int numToRead) {
+    char buff[REPLY_BUFF_SIZE];
+    RespReaderCtx respCtx;
+    readRespInit(&respCtx);
+
+    int retries = 0;
+    int delayUs = INITIAL_DELAY_USEC;
+
+    while ((int)respCtx.countReplies < numToRead) {
+        int rd = recv(ctx->fd, buff, sizeof(buff), MSG_DONTWAIT);
+
+        if (rd > 0) {
+            /* Data was received, process it */
+            if (RESP_REPLY_ERR == readRespReplies(&respCtx, buff, rd)) {
+                RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP_READ, respCtx.errorMsg);
+                return 1;
             }
-        } else {
-            //
-            RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_FAILED_READ,
-                            "Failed to read responses from socket.");
-            ctx->pendingCmds.num -= numRead;
+
+        } else if (rd == 0) {
+            RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_CONN_CLOSE, "Connection closed by the remote side");
             return 1;
+        } else {
+            /* Error occurred or no data available */
+            if (errno == EWOULDBLOCK || errno == EAGAIN) {
+                /* No data available, retry after delay */
+                if (retries >= MAX_RETRIES) {
+                    RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_MAX_RETRIES, "Maximum number of retries reached. Exit.");
+                    return 1;
+                }
+                usleep(delayUs);
+                delayUs *= 2;  /* Double the delay for each retry */
+                retries++;
+            } else {
+                RDB_reportError(ctx->p, (RdbRes) RDBX_ERR_RESP2TCP_FAILED_READ, "Failed to recv() from Redis. Exit.");
+                return 1;
+            }
         }
     }
-    ctx->pendingCmds.num -= numRead;
+
+    ctx->protocol.nResponses += respCtx.countReplies;
+    ctx->pendingCmds.num -= respCtx.countReplies;
     return 0;
 }
 
-/* return 0 for success. 1 Otherwise. */
-int tcpLoaderFlush(void *context) {
-    RdbxRespToTcpLoader *ctx = context;
-    if (ctx->pendingCmds.num)
-        return readReplies(ctx, ctx->pendingCmds.num);
-    return 0;
+static inline void recordNewCmd(RdbxRespToTcpLoader *ctx, const struct iovec *iov, int iovcnt) {
```

Comment:
maybe i'm not devoting enough time to read the code to understand it, but either way, i guess it would be better to add a 2 line comment on each function as to what it does, and if needed also document the inputs and outputs.

additionally, in there's some multi-function mechanism / design of command recording, maybe there should be some big doc block at the top of the file that describes it.

Reply:
Right now it only records. I will need to devote another phase for debugging and testing error flows.